Add Ritz options (largest/smallest in real/imaginary parts) for compl…

wztzjhn opened the pull request:
This pull request changes only the file "arpack.hpp"

## Pull request purpose

For complex eigenvalue routines, the Ritz value options "LR", "SR", "LI", "SI" were not provided in the C++ binding. Now they are included.
For details, please refer to the book "ARPACK Users' Guide" page 105, Table A.9

sylvestre commented:
would it be possible to add tests?
 thanks

wztzjhn commented:
Hi, what is the standard folder/file to add tests here? Did you mean tests for CI purpose?

sylvestre commented:
here:
https://github.com/opencollab/arpack-ng/blob/master/TESTS/icb_arpack_cpp.cpp

wztzjhn commented:
Thanks! I will take a look and come back

wztzjhn commented:
Question: 
I tried "make check", and went inside folder "TESTS". There are two executables generated that I saw: "bug1312" and "dnsimp". It seems the file "icb_arpack_cpp.cpp" was not compiled to start with. Any idea to where to change the behavior so I can add more stuff inside?

FYI, I compiled the library through the following (on Ubuntu 22.04):
FFLAGS="-m64 -I${MKL_INC_DIR}" FCFLAGS="-m64 -I$MKL_INC_DIR" LIBS="-Wl,--start-group ${MKL_LIB_DIR}/libmkl_gf_ilp64.a ${MKL_LIB_DIR}/libmkl_sequential.a ${MKL_LIB_DIR}/libmkl_core.a -Wl,--end-group -lpthread -lm -ldl" LIBSUFFIX="ILP64" INTERFACE64="1" ./configure --with-blas=mkl_gf_ilp64 --with-lapack=mkl_gf_ilp64 --enable-icb

wztzjhn commented:
Actually, there seems to be some problem when "make check". The two executables I mentioned above did run without emitting obvious mistakes. However, I notice some error messages when using "make check":

Making check in .
make[1]: Entering directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng'
make[1]: Nothing to be done for 'check-am'.
make[1]: Leaving directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng'
Making check in ICB
make[1]: Entering directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/ICB'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/ICB'
Making check in UTIL
make[1]: Entering directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/UTIL'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/UTIL'
Making check in SRC
make[1]: Entering directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/SRC'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/SRC'
Making check in TESTS
make[1]: Entering directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/TESTS'
make  dnsimp bug_1323 bug_58_double bug_79_double_complex bug_142 bug_142_gen icb_arpack_c icb_arpack_cpp bug_1315_single bug_1315_double testA.mtx
make[2]: Entering directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/TESTS'
f77  -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8 -c -o dnsimp.o dnsimp.f
f77  -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8 -c -o mmio.o mmio.f
/bin/bash ../libtool  --tag=F77   --mode=link f77  -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8   -o dnsimp dnsimp.o mmio.o ../SRC/libarpackILP64.la   -Wl,--start-group /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_gf_ilp64.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_sequential.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_core.a -Wl,--end-group -lpthread -lm -ldl
libtool: link: f77 -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8 -o .libs/dnsimp dnsimp.o mmio.o -Wl,--start-group -Wl,--end-group  ../SRC/.libs/libarpackILP64.so /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_gf_ilp64.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_sequential.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_core.a -lpthread -lm -ldl -Wl,-rpath -Wl,/home/ztwang/installs/lib
f77  -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8 -c -o bug_1323.o bug_1323.f
/bin/bash ../libtool  --tag=F77   --mode=link f77  -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8   -o bug_1323 bug_1323.o ../SRC/libarpackILP64.la   -Wl,--start-group /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_gf_ilp64.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_sequential.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_core.a -Wl,--end-group -lpthread -lm -ldl
libtool: link: f77 -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8 -o .libs/bug_1323 bug_1323.o -Wl,--start-group -Wl,--end-group  ../SRC/.libs/libarpackILP64.so /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_gf_ilp64.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_sequential.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_core.a -lpthread -lm -ldl -Wl,-rpath -Wl,/home/ztwang/installs/lib
f77  -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8 -c -o bug_58_double.o bug_58_double.f
/bin/bash ../libtool  --tag=F77   --mode=link f77  -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8   -o bug_58_double bug_58_double.o ../SRC/libarpackILP64.la   -Wl,--start-group /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_gf_ilp64.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_sequential.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_core.a -Wl,--end-group -lpthread -lm -ldl
libtool: link: f77 -m64 -I/opt/intel/oneapi/mkl/2022.1.0/include -fdefault-integer-8 -o .libs/bug_58_double bug_58_double.o -Wl,--start-group -Wl,--end-group  ../SRC/.libs/libarpackILP64.so /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_gf_ilp64.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_sequential.a /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_core.a -lpthread -lm -ldl -Wl,-rpath -Wl,/home/ztwang/installs/lib
/usr/bin/ld: /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_core.a(dgetrs.o): in function `mkl_lapack_dgetrs':
dgetrs_gen.f:(.text+0x224): undefined reference to `mkl_blas_dtrsm'
/usr/bin/ld: dgetrs_gen.f:(.text+0x2b1): undefined reference to `mkl_blas_dtrsm'
/usr/bin/ld: dgetrs_gen.f:(.text+0x2e7): undefined reference to `mkl_lapack_dlaswp'
/usr/bin/ld: dgetrs_gen.f:(.text+0x324): undefined reference to `mkl_lapack_dlaswp'
/usr/bin/ld: dgetrs_gen.f:(.text+0x3b0): undefined reference to `mkl_blas_dtrsm'
/usr/bin/ld: dgetrs_gen.f:(.text+0x43f): undefined reference to `mkl_blas_dtrsm'
/usr/bin/ld: dgetrs_gen.f:(.text+0x4ae): undefined reference to `mkl_blas_dtrsv'
/usr/bin/ld: dgetrs_gen.f:(.text+0x513): undefined reference to `mkl_blas_dtrsv'
/usr/bin/ld: dgetrs_gen.f:(.text+0x571): undefined reference to `mkl_blas_dtrsv'
/usr/bin/ld: dgetrs_gen.f:(.text+0x5d5): undefined reference to `mkl_blas_dtrsv'
/usr/bin/ld: /opt/intel/oneapi/mkl/2022.1.0/lib/intel64/libmkl_core.a(dgetrf_local.o): in function `mkl_lapack_dgetrf_local':
dgetrf_local_gen.f:(.text+0x36e): undefined reference to `mkl_lapack_dlaswp'
/usr/bin/ld: dgetrf_local_gen.f:(.text+0x432): undefined reference to `mkl_blas_dtrsm'
/usr/bin/ld: dgetrf_local_gen.f:(.text+0x685): undefined reference to `mkl_lapack_dlaswp'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:729: bug_58_double] Error 1
make[2]: Leaving directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/TESTS'
make[1]: *** [Makefile:1188: check-am] Error 2
make[1]: Leaving directory '/home/ztwang/Documents/Code_Projects/GitHub/arpack-ng/TESTS'
make: *** [Makefile:531: check-recursive] Error 1

fghoussen commented:
@wztzjhn: can you update the change log?

wztzjhn commented:
Yes, happy to. I see the file "CHANGES" starting like the follow, and should I put my changes above/below "arpack-ng - 3.9.0"? 

----------
arpack-ng - 3.9.0

[ Jose E. Roman ]
* Avoid using isnan() in tests, since is GNU-specific

[ Tom Payerle ]
* Change the continuation line format for stat.h, debug.h

fghoussen commented:
Below. Current release is 3.8.0.

wztzjhn commented:
@fghoussen  Thanks! updated.

wztzjhn commented:
I have added a few more changes to the test file "icb_arpack_cpp.cpp":

1. change "ldz" to the minimal allowed value "N", "z" to the minimal allowed size "ldz * nev", and "d" to the minimal allowed size "nev";

2. change "lworkl" in the real symmetric runner to the correct minimal value "ncv * (ncv + 8)" (for reference, see SRC/ssaupd.f, SRC/dsaupd.f);

3. change "lworkl" in the complex nonsymmetric runner to the correct minimal value "ncv * (3 * ncv + 5)" (for reference, see SRC/cnaupd.f, SRC/znaupd.f);

4. remove initialization of the array "select", since it is only a workspace in the current context (initial value not used);

5. change size of the "ipntr" array in the real symmetric runner from 14 to the minimal allowed value "11" (for reference, see SRC/ssaupd.f, SRC/dsaupd.f);

6. remove initialization of iparam[4] -- it is only an output;

7. change the reverse communication condition "while (ido != 99)" to the more specific "while (ido == 1 || ido == -1)" for the standard eigenvalue problems used in the test;

8. throw exception when info < 0;

9. add and removed a few comments for clarity.

wztzjhn commented:
Same changes in file icb_parack_c.c added.

fghoussen commented on TESTS/icb_arpack_c.c:
```diff
@@ -38,37 +38,35 @@ int ds() {
-  iparam[0] = 1;
-  iparam[2] = 10 * N;
-  iparam[3] = 1;
-  iparam[4] = 0;  // number of ev found by arpack.
```
Can you let `iparam[4]` initialized to `0`?
Initialisation is always good thing even if it's an output.

wztzjhn replied:
For the initialization discussion: I am neutral in choosing to initialize (or not) the output-only arguments, but inclined to have a consistent treatment of all these arguments in the same function. If we choose to initialize iparam[4] because "initialization is always a good thing", by the same token we should also initialize all other elements in "iparam[0:10]", and also all elements in "ipntr[:]", and also all elements in the eigenvalues -- the list simply goes on... So I feel like just not necessary to do so just particularly for iparam[4], since it is not special at all compared to other variables mentioned above.

What do you think? If you want, I can still switch on ipram[4] initialization...

fghoussen commented on TESTS/icb_arpack_c.c:
```diff
@@ -109,38 +107,36 @@ int zn() {
-  iparam[0] = 1;
-  iparam[2] = 10 * N;
-  iparam[3] = 1;
-  iparam[4] = 0;  // number of ev found by arpack.
```
Can you let iparam[4] initialized to 0?
Initialisation is always good thing even if it's an output.

wztzjhn replied:
For the initialization discussion: I am neutral in choosing to initialize (or not) the output-only arguments, but inclined to have a consistent treatment of all these arguments in the same function. If we choose to initialize iparam[4] because "initialization is always a good thing", by the same token we should also initialize all other elements in "iparam[0:10]", and also all elements in "ipntr[:]", and also all elements in the eigenvalues -- the list simply goes on... So I feel like just not necessary to do so just particularly for iparam[4], since it is not special at all compared to other variables mentioned above.

What do you think? If you want, I can still switch on ipram[4] initialization...

fghoussen commented on TESTS/icb_arpack_cpp.cpp:
```diff
@@ -26,50 +25,48 @@ void diagonal_matrix_vector_product(Real const* const x, Real* const y) {
-  iparam[0] = 1;
-  iparam[2] = 10 * N;
-  iparam[3] = 1;
-  iparam[4] = 0;  // number of ev found by arpack.
```
Can you let iparam[4] initialized to 0?
Initialisation is always good thing even if it's an output.

fghoussen commented on TESTS/icb_arpack_cpp.cpp:
```diff
@@ -98,63 +93,64 @@ void real_symmetric_runner(double const& tol_check) {
-  iparam[0] = 1;
-  iparam[2] = 10 * N;
-  iparam[3] = 1;
-  iparam[4] = 0;  // number of ev found by arpack.
```
Can you let iparam[4] initialized to 0?
Initialisation is always good thing even if it's an output.

wztzjhn replied:
For the initialization discussion: I am neutral in choosing to initialize (or not) the output-only arguments, but inclined to have a consistent treatment of all these arguments in the same function. If we choose to initialize iparam[4] because "initialization is always a good thing", by the same token we should also initialize all other elements in "iparam[0:10]", and also all elements in "ipntr[:]", and also all elements in the eigenvalues -- the list simply goes on... So I feel like just not necessary to do so just particularly for iparam[4], since it is not special at all compared to other variables mentioned above.

What do you think? If you want, I can still switch on ipram[4] initialization...

fghoussen commented:
> Same changes in file icb_parack_c.c added.

You meant `icb_arpack_c.c`? `icb_parpack_c.c` is unchanged.

wztzjhn commented:
> > Same changes in file icb_parack_c.c added.
> 
> You meant `icb_arpack_c.c`? `icb_parpack_c.c` is unchanged.

You are right -- I meant "icb_arpack_c.c" not the parpack file.

For the initialization discussion: I am neutral in choosing to initialize (or not) the output-only arguments, but inclined to have a consistent treatment of all these arguments in the same function. If we choose to initialize iparam[4] because "initialization is always a good thing", by the same token we should also initialize all other elements in "iparam[0:10]", and also all elements in "ipntr[:]", and also all elements in the eigenvalues -- the list simply goes on... So I feel like just not necessary to do so just particularly for iparam[4], since it is not special at all compared to other variables mentioned above.

What do you think? If you want, I can still switch on ipram[4] initialization...

fghoussen commented:
Can you do the same kind of clean-up for `parpack.hpp`, `icb_parpack_c.c`, `icb_parpack_cpp.cpp`? Thanks!

> What do you think? If you want, I can still switch on ipram[4] initialization...

OK, I give you that :)

wztzjhn commented:
> Can you do the same kind of clean-up for `parpack.hpp`, `icb_parpack_c.c`, `icb_parpack_cpp.cpp`? Thanks!
> 
> > What do you think? If you want, I can still switch on ipram[4] initialization...
> 
> OK, I give you that :)

Thanks! Will do the same for the files you mentioned

wztzjhn commented:
@fghoussen 
Hi, previously I disabled MPI so the check works. Now, with MPI enabled to work on parpack, I am hitting a failure in the check, which is traced back to an old commit **aa970ce** with commit message "**ILP64 support. (#165)**", on file "PARPACK/TESTS/MPI/issue46.f". Here is what I see and what I understand:
--- what I see ---
![Screenshot from 2022-08-07 17-59-19](https://user-images.githubusercontent.com/5083567/183314335-99132f2d-c560-4b82-9ae1-727aa3e45e41.png)

--- what I standstand ---
The file "issue46.f" started with "**integer** ierr, comm, color, key, myid, nprocs, cnprocs", until the commit **aa970ce** when it becomes "**integer*4** ierr, comm, myid, nprocs, cnprocs, color" where the commit message points to fixing "ILP64 support". 

Now if I look at the discrepancy, "MPI_COMM_WORLD", defined in file "/usr/lib/x86_64-linux-gnu/fortran/gfortran-mod-15/openmpi/mpif-handles.h" line 42 (I am using Ubuntu 22.04 openmpi 4.1.2, and ILP64 is enabled), it says "**integer** MPI_COMM_WORLD", which explained the error message.

I appreciate your great amount of work in providing the ILP64 support, but also I am a bit scared away trying to fix the issue here: looking at the long commit history related to ILP64 support, I don't think I have enough knowledge/time here to fix the issue at face value without creating newer ones. Would it be possible for the original authors (I see only 3 related to the file "issue46.f") to first fix those issues? I am still happy to do my part on cleaning up the parpack test files once I can pass the check.

FYI, here is my environment for reproducing the issues:
system: PopOS 22.04 (practically identical to Ubuntu 22.04)
MKL: newest, directly downloaded from INTEL website
MPI: openmpi 4.1.2
gfortran: 11.2.0

The arpack-ng installation with autotools:
export MKL_INC_DIR=${MKLROOT}/include
export MKL_LIB_DIR=${MKLROOT}/lib/intel64

FFLAGS="-m64 -I${MKL_INC_DIR}" FCFLAGS="-m64 -I$MKL_INC_DIR" CFLAGS="-DMKL_ILP64 -m64 -I${MKL_INC_DIR}" CXXFLAGS="-DMKL_ILP64 -m64 -I${MKL_INC_DIR}" LIBS="-L${MKL_LIB_DIR} -Wl,--no-as-needed -lmkl_gf_ilp64 -lmkl_tbb_thread -lmkl_core -lpthread -ltbb -lstdc++ -lm -ldl" LIBSUFFIX="ILP64" INTERFACE64="1" ./configure --with-blas=mkl_gf_ilp64 --with-lapack=mkl_gf_ilp64 --enable-mpi --enable-icb --prefix=$HOME/installs

wztzjhn commented:
Hmm, looks like the current neighboring pull request is also talking about "issue46.f"? :P

fghoussen commented:
@wztzjhn: do not use `INTERFACE64=1` ! Running with MPI on any distro _without_ `INTERFACE64` and any blas/lapack (netlib, MKL, ...) should be OK. :)

Support for ILP64 arpack was a hell... But should be OK now...
Support for ILP64 for **p**arpack (MPI version of arpack) will be even worse... If not _impossible_: look at 600+ commits here #368... Don't set `INTERFACE64`!

wztzjhn commented:
OK. Will give a try without ILP64. Did not realize that ILP64 does not work with MPI yet...

Just on a different thought: what will be the use case of parallel arpack? In my past experience, the serial arpack actually works well for matrices of dimension ~ 10^7 even up to 10^8. Beyond that point I could see a value of using MPI, also at the same time ILP64 becomes necessary. So in my view MPI and ILP64 really come hand in hand -- it is indeed a pity that the work of combining MPI and ILP64 is such a beast...

wztzjhn commented:
@fghoussen 
Mind if I change expression like "double const x" to "const double x" in these test files? The latter format seems way more commonly adopted.

wztzjhn commented:
Please check, while working on the cleaning, I found two bugs (only one fixed as stated in the commit message). 

There is one remaining bug (not due to my changes) in the MPI complex (single precision) runner. 

To reproduce from the current master branch: change line 113 of file "icb_parpack_cpp.cpp" from "nev=1" to any number bigger than 1. You will see that this single precision complex runner produces multiple (equal to the number of MPI threads) degenerate eigenvalues, while the input matrix has no degeneracy.

This bug did not show up in the normal tests, simply because this bug was known, and somebody changed nev=3 (or nev=9) as in other test files to nev=1, so people look away from the bug. But I am afraid as time goes by people may forget about it and assume it is correct...

To see it more clearly, you can check out my branch, where I have made a template to switch the precision. The latest commit switched the precision from single to double (line 212), which produces the correct results. Once you change that line to "float", it goes back to the old buggy behavior.

wztzjhn commented:
Looked into pc/znaupd.f, did not find any clear clue.

fghoussen commented:
If #372 runs OK (green CI), can you kill [e2f0483](https://github.com/opencollab/arpack-ng/pull/371/commits/e2f04839b00a6dc5a19028f500b5fd1a931d9b7f) and rebase on top of it? 

> what will be the use case of parallel arpack? In my past experience, the serial arpack actually works well for matrices of dimension ~ 10^7 even up to 10^8.

Indeed, arpack can handle high dimensions (because matrices are sparse). AFAIK, parpak can be used to go higher dimension than arpack (but in practice, it's rare you need that because arpack can go high already), or, go faster than arpack for the same (high) dimension matrices.

> it is indeed a pity that the work of combining MPI and ILP64 is such a beast...

Unfortunately, I guess #368 will be a dead end (I am about to close it). I have still hope on this but no time...

> Mind if I change expression like "double const x" to "const double x" in these test files? 

OK

> Please check, while working on the cleaning, 

I'll have a look when possible (week-end probably)

> There is one remaining bug (not due to my changes) in the MPI complex (single precision) runner.  Had a quick look at pc/znaupd.f, did not find any clear clue.

Thanks! Good shot: likely connected to #332. 
I cherry-picked [e2f0483](https://github.com/opencollab/arpack-ng/pull/371/commits/e2f04839b00a6dc5a19028f500b5fd1a931d9b7f) in #372 so you can remove it from here.
Clue is likely in between `parpack.h` and `pc/znaupd.f,`: checkout [ae98874](https://github.com/opencollab/arpack-ng/pull/372/commits/ae9887494991e4f6aac366ba25eadd56ead1c498) in #372 

> To see it more clearly, you can check out my branch, where I have made a template to switch the precision. The latest commit switched the precision from single to double (line 212), which produces the correct results. Once you change that line to "float", it goes back to the old buggy behavior.

If #372 runs OK (green CI), keep both tests in float and double.

wztzjhn commented:
> If #372 runs OK (green CI), can you kill [e2f0483](https://github.com/opencollab/arpack-ng/pull/371/commits/e2f04839b00a6dc5a19028f500b5fd1a931d9b7f) and rebase on top of it?

Sure. Just making sure of the operations: 
1. #372 becomes green and is merged into master.
2. I add a new branch and cherry pick everything except that bug fix.
3. Simply close this pull request, then create a new pull request from the new branch.
Is that what you have in mind?

fghoussen commented:
>1. https://github.com/opencollab/arpack-ng/pull/372 becomes green and is merged into master.

Yes.

>2. I add a new branch and cherry pick everything except that bug fix.

Simply `git rebase -i master` and get rid of commits already merged into master (your bug fix).

>Simply close this pull request, then create a new pull request from the new branch. Is that what you have in mind?

Push-force and this PR will follow master.

wztzjhn commented:
> > 1. [Fix issue 332 #372](https://github.com/opencollab/arpack-ng/pull/372) becomes green and is merged into master.
> 
> Yes.
> 
> > 2. I add a new branch and cherry pick everything except that bug fix.
> 
> Simply `git rebase -i master` and get rid of commits already merged into master (your bug fix).
> 
> > Simply close this pull request, then create a new pull request from the new branch. Is that what you have in mind?
> 
> Push-force and this PR will follow master.

Thanks! Will do.

fghoussen commented:
Can you rebase? (1-change-1-commit if possible: easier to review)

wztzjhn commented:
> Can you rebase? (1-change-1-commit if possible: easier to review)

Done. At this moment all CI's seem to be working, except that the macos is slowly installing the dependencies. Hopefully they will be green.

wztzjhn commented:
Just a minor digression: for macos it will be much faster to cache the dependencies in CI. I don't know if it is easy to do the cache with homebrew (since it has a default installation location which I don't know if the workflow job can access); An alternative option is to use mini-conda instead of homebrew, create a new environment and set your desired library location, then cache the library there.

fghoussen commented on TESTS/icb_arpack_cpp.cpp:
```diff
@@ -94,106 +82,109 @@ void real_symmetric_runner(double const& tol_check) {
-  real_symmetric_runner<double>(1.e-05);
+  real_symmetric_runner<float>(1., arpack::which::largest_magnitude);
+  real_symmetric_runner<double>(1.e-05, arpack::which::largest_magnitude);
 
```
Possible to add 2 tests (float/double) with `arpack::which::largest_real` for instance? If not, no big deal.

wztzjhn replied:
not a problem, will do

fghoussen commented:
>  for macos it will be much faster to cache the dependencies in CI. 

Sure. Caching was used with (old) travis-ci but not with (new) github-actions.
No time to look at it. Not sure if system files `\usr\*` could be cached: try it here #373 .

fghoussen commented:
Fixing #312

wztzjhn commented on ICB/parpack.h:
```diff
@@ -14,16 +14,21 @@
 extern "C" {
```
Here it is literally exchanging a few function-declaration lines, so only for clarity but should not alter any result. But if you  want I will revert them here.

wztzjhn commented on PARPACK/SRC/MPI/pcnaupd.f:
```diff
@@ -54,11 +54,11 @@
 c
```
Here it is literally exchanging a few function-declaration lines, so only for clarity but should not alter any result. But if you want I will revert them here.

wztzjhn commented on PARPACK/SRC/MPI/pznaupd.f:
```diff
@@ -9,14 +9,14 @@
 c  iteration. This is intended to be used to find a few eigenpairs of a
```
Here it is literally exchanging a few function-declaration lines, so only for clarity but should not alter any result. But if you want I will revert them here.

fghoussen replied:
No big deal: I'll handle that before merging.

wztzjhn commented on PARPACK/TESTS/MPI/icb_parpack_cpp.cpp:
```diff
@@ -231,7 +209,7 @@ int main() {
 
   try {
-    complex_symmetric_runner();
+    complex_nonsymmetric_runner<float>();
```
In principle yes, but I am not familiar with the debug routines there. So if we want it I may do it in a week (busy recently); or if you are already familiar with the debug routines probably you can do it in a minute?

fghoussen replied:
OK, I'll do that

wztzjhn replied:
Thanks!
